cunumeric/module.py

```diff
@@ -6227,3 +6227,688 @@ def bincount(
+    q_arr,
+    axis,
+    axes_set,
+    original_shape,
+    method,
+    keepdims,
+    to_dtype,
+    qs_all,
+):
+
+    ndims = len(arr.shape)
+
+    if axis is None:
+        n = arr.size
+        remaining_shape = []  # only `q_arr` dictates shape;
+        # quantile applied to `arr` seen as 1D;
+    else:
+        n = arr.shape[axis]
+
+        # arr.shape -{axis}; if keepdims use 1 for arr.shape[axis]:
+        # (can be empty [])
+        #
+        if keepdims:
+            remaining_shape = [
+                1 if k in axes_set else original_shape[k]
+                for k in range(0, len(original_shape))
+            ]
+        else:
+            remaining_shape = [
+                arr.shape[k] for k in range(0, ndims) if k != axis
+            ]
+
+    # quantile interpolation method to act on vectors,
+    # like an axpy call:
+    # [gamma](vleft, vright) { return (1-gamma)*vleft + gamma*vright;}
+    #
+    linear_interp = (
+        lambda gamma, arr_lvals, arr_rvals: (1.0 - gamma) * arr_lvals
+        + gamma * arr_rvals
+    )
+
+    # compose qarr.shape with arr.shape:
+    #
+    # result.shape = (q_arr.shape, arr.shape -{axis}):
+    #
+    qresult_shape = (*q_arr.shape, *remaining_shape)
+
+    # fails with cunumeric:
+    # (numpy works)
+    #
+    # cunumeric BUG: TypeError: a bytes-like object is required, not 'ndarray';
+    # TODO: open nvbug
+    #
+    # qs_all = ndarray(shape = qresult_shape,
+    #                     buffer = array(result_1D),
+    #                     dtype = to_dtype)
+    #
+    # construct result NdArray, non-flattening approach:
+    #
+    if qs_all is None:
+        qs_all = zeros(qresult_shape, dtype=to_dtype)
+    else:
+        # implicit conversion from to_dtype to qs_all.dtype assumed
+        #
+        if qs_all.shape != qresult_shape:
+            raise ValueError("wrong shape on output array")
+
+    for index, q in np.ndenumerate(q_arr):
+        (gamma, j) = method(q, n)
+        (left_pos, right_pos) = (j, j + 1)
+
+        # (N-1) dimensional ndarray of left, right
+        # neighbor values:
+        #
+        # non-flattening approach:
+        #
+        # extract values at index=left_pos;
+        arr_1D_lvals = arr.take(left_pos, axis)
+        arr_vals_shape = arr_1D_lvals.shape
+
+        if right_pos >= n:
+            # some quantile methods may result in j==(n-1),
+            # hence (j+1) could surpass array boundary;
+            #
+            arr_1D_rvals = zeros(arr_vals_shape, dtype=arr_1D_lvals.dtype)
+        else:
+            # extract values at index=right_pos;
+            arr_1D_rvals = arr.take(right_pos, axis)
+
+        # this is the main source of parallelism
+        # (the other one being on `q` inputs; ignored for now)
+        #
+        quantiles_ls = linear_interp(gamma, arr_1D_lvals, arr_1D_rvals)
+
+        # non-flattening approach:
+        #
+        if len(index) == 0:
+            qs_all[:] = quantiles_ls.reshape(qs_all.shape)
```

Comment:
I had to change this to `qs_all[...] =`, because `qs_all[:]` doesn't seem to work with 0-dim arrays (`:` matches 1+ dimensions, whereas `...` matches 0+ dimensions).

This happens e.g. in this example:
```
cn.quantile(cn.ones((3,4,5)), 0.5)
```

Reply:
Done.